pkg/client/apiutil/restmapper.go

```diff
@@ -194,13 +208,28 @@ func (m *mapper) addKnownGroupAndReload(groupName string, versions ...string) er
 		return fmt.Errorf("failed to get API group resources: %w", err)
 	}
 
-	if _, ok := m.knownGroups[groupName]; ok {
-		groupResources = m.knownGroups[groupName]
-	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.addGroupVersionResourcesToCache(groupVersionResources)
+	return nil
+}
 
+// addGroupVersionResourcesToCache does what the name suggests. The mutex must be held when
+// calling it.
+func (m *mapper) addGroupVersionResourcesToCache(gvr map[schema.GroupVersion]*metav1.APIResourceList) {
```

Comment:
```suggestion
func (m *mapper) addGroupVersionResourcesToCacheLocked(gvr map[schema.GroupVersion]*metav1.APIResourceList) {
```
nit

Just to be a bit safer going forward. Can you please go through all funcs in this file and add a Locked suffix to all that assume that a lock is held? (a few more example are: isGroupVersionCached, isAPIGroupCached)

I think it should be at least all funcs that are accessing m.knownGroups or m.apiGroups without locking themselves

Reply:
This was a good call. Updated things as suggested but also changed `addKnownGroupAndReload` to hold the lock everywhere rather than just when its accessing data stuctures that need locking. The reason is that without that, its possible andlikely during startup,that we might have multiple calls to `findAPIGroupByNameAndMaybeAggregatedDiscoveryLocked` going on in parallel, as the previously-unlocked api calls take much longer than updating the locked data structures. 

As `fetchGroupVersionResources` already required us to hold the lock there is no case where things might get worse.

Reply:
Sounds good, thx!